apiserver/facades/client/secrets/secrets.go

```diff
@@ -100,25 +97,35 @@ func (s *SecretsAPI) ListSecrets(ctx context.Context, arg params.ListSecretsArgs
 	if arg.Filter.Label != nil {
 		labels = append(labels, *arg.Filter.Label)
 	}
-	if arg.Filter.Revision != nil {
-		revisions = append(revisions, *arg.Filter.Revision)
-	}
 	if arg.Filter.OwnerTag != nil {
 		tag, err := names.ParseTag(*arg.Filter.OwnerTag)
 		if err != nil {
 			return params.ListSecretResults{}, errors.Trace(err)
 		}
 		switch kind := tag.Kind(); kind {
 		case names.ApplicationTagKind:
-			appOwners = append(appOwners, tag.Id())
+			owner = &secretservice.CharmSecretOwner{
+				Kind: secretservice.ApplicationOwner,
+				ID:   tag.Id(),
+			}
 		case names.UnitTagKind:
-			unitOwners = append(unitOwners, tag.Id())
+			owner = &secretservice.CharmSecretOwner{
+				Kind: secretservice.UnitOwner,
+				ID:   tag.Id(),
+			}
 		default:
 			return result, errors.NotValidf("secret owner tag kind %q", kind)
 		}
 	}
-	wantUser = len(appOwners) == 0 && len(unitOwners) == 0
-	metadata, revisionMetadata, err := s.secretService.ListSecrets(ctx, uri, revisions, labels, appOwners, unitOwners, wantUser)
+	var (
+		metadata         []*coresecrets.SecretMetadata
+		revisionMetadata [][]*coresecrets.SecretRevisionMetadata
+	)
+	if owner == nil {
+		metadata, revisionMetadata, err = s.secretService.ListSecrets(ctx, uri, arg.Filter.Revision, labels)
+	} else {
+		metadata, revisionMetadata, err = s.secretService.ListCharmSecrets(ctx, *owner)
```

Comment:
The fallback was always there because that's how the mongo state stuff worked under the hood. It was done in one ListSecrets() querying using the supplied args. Because we now have structured the secrets service method to have different List methods for user vs charm secrets, the logic has bubbled up to here. I'll update the comment.